### PR TITLE
Pokemon

### PR DESCRIPTION
**Pull request by NikolaJelic**
This PR is referencing issue #1. 

The Pokemon loads it's `next_move` which is selected by the player from a list of possible options. The move selection and the possible moves, fetched from the Pokemon API will be implemented by @hentai-chan .
The attack damage is calculated using the formula found  [here](https://www.math.miami.edu/~jam/azure/compendium/battdam.htm
)

# Testing and usage
```cpp

	Move stab = Move("stab", MoveTypes::ATTACK, 80, 30);
	Move heal = Move("heal", MoveTypes::HEAL, 100, 50);

	Pokemon bulbasaur = Pokemon("Bulbasaur", 10, 50, 30, stab);
	Pokemon pikatchu = Pokemon("Pikatchu", 3, 50, 30, heal);
	
	std::cout << "Pikatchu hp:" << pikatchu.m_hp << std::endl;	
	std::cout << " Bulbasaur used: " << bulbasaur.make_move(bulbasaur, pikatchu).m_move_name << std::endl;
	std::cout << "Pikatchu hp after 1. attack:" << pikatchu.m_hp << std::endl;
	pikatchu.make_move(pikatchu, pikatchu);
	std::cout << "Pikatchu hp after heal:" << pikatchu.m_hp << std::endl;
```
A move can only be made if the Pokemon `isAlive = true`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -35,6 +35,9 @@ out/
 
 # Visual Studio 2015/2017 cache/options directory
 .vs/
+**/.vs/*
+**/.vscode/*
+.vscode/
```
Comment (karnkaul) [inline .gitignore]: This is redundant

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (karnkaul) [inline .vscode/launch.json]: This file should not be committed

File: .vscode/tasks.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (karnkaul) [inline .vscode/tasks.json]: This file should not be committed

File: CMakeLists.txt [hunk 4]
```diff
@@ -8,6 +8,7 @@ set(APP_VERSION "1.0.0" CACHE STRING "Application Version" FORCE)
 project(pkmn VERSION ${APP_VERSION})
 
 file(GLOB_RECURSE SOURCES CONFIGURE_DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/src/*.*pp")
+message("sources: ${SOURCES}")
```
Comment (karnkaul) [inline CMakeLists.txt]: This looks like debugging code, remove it

File: src/main.cpp [hunk 5]
```diff
@@ -1,11 +1,12 @@
...
 #include <string>
+#include "models.hpp"
 #include "utils.hpp"
-
 int main()
 {
-	auto name = get_user_input<std::string>("What's your name? ");
-	std::cout << "Hello, " << name << '!' << '\n';
-	std::cin.get();
+	/* 	auto name = get_user_input<std::string>("What's your name? ");
```
Comment (karnkaul) [inline src/main.cpp:7]: So, the program just returns 0?

File: src/models.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
+#include "models.hpp"
+#include <iostream>
+#include <vector>
+#include "utils.hpp"
+
+Move::Move(std::string move_name, MoveTypes mt, float accuracy, int power)
+{
+	m_move_name = move_name;
```
Comment (karnkaul) [inline src/models.cpp]: Member initializer list / rule of zero

File: src/models.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+Move Pokemon::make_move(Pokemon& source, Pokemon& target)
+{
+	if ((random_range<int>(0, 100) <= source.m_next_move.m_accuracy) && source.isAlive && target.isAlive)
+	{
+		switch (source.m_next_move.m_mt)
+		{
+		case MoveTypes::ATTACK:
+			return attack(source, target);
+			break;
```
Comment (karnkaul) [inline src/models.cpp]: Redundant break

File: src/models.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+			return source.m_next_move;
+		}
+	}
+	Move fail = Move("Failed", MoveTypes::NONE, 100, 0);
+	return fail;
+};
+Move Pokemon::attack(Pokemon& source, Pokemon& target)
+{
+	/* (2Level/5+2)*Attack*Power)/Defense)/50)+2)*random number(217-255))/255
+			damage formula https://www.math.miami.edu/~jam/azure/compendium/battdam.htm*/
```
Comment (karnkaul) [inline src/models.cpp]: Spaces required between identifiers in the maths expression

File: src/models.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+		target.m_hp -= damage;
+		if (target.m_hp <= 0)
+		{
+			target.isAlive = false;
+		}
+		return source.m_next_move;
+	}
+	return Move("Failed", MoveTypes::NONE, 100, 0);
+};
+Move Pokemon::defend(Pokemon& source)
```
Comment (karnkaul) [inline src/models.cpp:57]: Might want some whitespace between function definitions, and remove semicolons after function definitions

File: src/models.hpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+	std::string m_move_name;
+	MoveTypes m_mt;
+	float m_accuracy;
+	int m_power;
+	Move(std::string move_name, MoveTypes mt, float accuracy, int power);
+};
+
+struct Pokemon
+{
+	std::string m_name;
```
Comment (karnkaul) [inline src/models.hpp:27]: This struct is too huge and flat, use sub structs to improve it

File: src/models.hpp [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+	std::int32_t m_def{};
+	std::int32_t m_attack{};
+	bool isAlive = true;
+
+	std::vector<Move> m_moveset;
+	Move m_next_move = Move("None", MoveTypes::NONE, 100, 0);
+
+	Pokemon() = default; // default constructor can be compiler generated since you are using default member initialization
+	Pokemon(std::string Name, int level, int def, int attack, Move next_move);
+	Move make_move(Pokemon& source, Pokemon& target);
```
Comment (karnkaul) [inline src/models.hpp:40]: Are none of these member functions const?

File: src/utils.hpp [hunk 12]
```diff
@@ -63,3 +64,4 @@ enum class Color
 };
 
 std::string style(std::string text, Color fore, Color back = Color::BLACK);
+int random_range(int min, int max);
```
Comment (karnkaul) [inline src/utils.hpp]: Why is this needed?

**Review by karnkaul — CHANGES_REQUESTED** [review]

**Review by karnkaul — APPROVED** [review]

**Review by karnkaul — APPROVED** [review]

File: src/models.hpp [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+struct Pokemon
+{
+	std::string m_name;
+
+	std::int32_t m_level = 1;
+	std::int32_t m_hp = 4 * m_level * random_range(0.8, 1.2);
+	std::int32_t m_max_hp = m_hp;
+	std::int32_t m_def{};
+	std::int32_t m_attack{};
+	bool isAlive = true;
```
Comment (hentai-chan) [inline src/models.hpp:34]: Still think checking against `if (hp > 0) { .. }` is just as readable but ok

Comment (karnkaul) [reply src/models.hpp]: I'll push back against having this field at all: it's unnecessary maintenance of state which is 100% derived from other, already existing state. Use a member function instead:
```cpp
constexpr bool is_alive() {
  return m_hp > 0;
}
```

File: src/models.cpp [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+	/* (2 * Level / 5+2) * Attack * Power) / Defense) / 50) + 2) * random number(217-255)) / 255
+			damage formula https://www.math.miami.edu/~jam/azure/compendium/battdam.htm*/
+	if (target.m_hp > 0)
+	{
+		int attack = source.m_attack;
+		if (random_range<int>(random_range<int>(1, 20), 100))
+		{
+			attack *= 2; // crit damage
+		}
+		int damage = ((((((2 * source.m_level / 5 + 2) * attack * source.m_next_move.m_power) / target.m_def) / 50) + 2) * random_range<int>(217, 255)) / 255;
```
[the remark this answers is not in the document]
Comment (NikolaJelic) [reply src/models.cpp:49]: It is linked at the start of that function

[the remark this answers is not in the document]
Comment (hentai-chan) [reply src/models.cpp:49]: Better put the comment directly above the line where this formula is used, makes for a better reading flow but that's just nitpicking :)

[the remark this answers is not in the document]
Comment (NikolaJelic) [reply src/models.cpp:49]: Done

**Review by hentai-chan — APPROVED** [review]
